### DISCUSSION TITLE
TEENY.M10 for "german" M10

### DISCUSSION THREAD

**Discussion by a-brandt** [bkw777/dlplus#8] [General]
Hello bkw777,

I am an owner of a german Olivetti M10. But the TEENY.M10 program works only on the USA version of the M10.

Do you know if there is a version of TEENY.M10 for the German version of a M10?

Best Regards
a-brandt

**Comment by bkw777** [comment]
Unfortunately I don't know of any tpdd dos for that combination. I have asked on the [m1000 mail list](http://lists.bitchin100.com/listinfo.cgi/m100-bitchin100.com)

Teeny has been ported and modded by a few people in the past to create versions like the US Olivetti in the first place, so it's theoretically possible, but it's beyond me.

Another possible source of info is recently someone made substantial changes to customize it and then build it into a custom main rom, and they supply an asm file that describes the changes they made, https://sarahkmarr.com/retromodel100.html but again, I don't know how to get from there to a working binary. If a binary existed, I could made a loader for it.

One thing is for sure, no one can even start without having a dump of the main rom to find all the special addresses and build a translation table. Do you have the ability to dump the main rom with an eprom programmer? 
If not, it may be possible to get you a small basic program that can read the main rom and dump it out the serial port as hex.

Actually, maybe the German rom is just the EU rom? We have an EU main rom for M10 here
http://tandy.wiki/Model_T_Y2K
So maybe we have a rom image that could be used to build a translation table.

**Comment by a-brandt** [comment]
Yes it is a EU rom.

The Olivetti M10 has different keyboards for different languages (german a has QWERTZ keyboard). But this a a setting on the motherboard.

**Comment by bkw777** [comment]
I found a starting point
https://github.com/LivingM100SIG/Living_M100SIG/blob/main/M100SIG/Lib-12-NEC-OLIVETTI/M10ROM.DIF

I can install the US teeny on my M10 to generate a plain binary .CO instead of the encoded payload in the .DO loader, and then maybe it's possible to walk that looking for the matching addresses to translate.

We could also diff a 100-us version with a M10-us version to see what parts actually need to be changed.

**Comment by majick** [comment]
Diffs between the M10-US and M100-US ROMs are also here:
https://github.com/z88dk/techdocs/blob/master/targets/m100/ROM_COMPARE

With an excellent level of detail in the according disassembly here:
https://github.com/z88dk/techdocs/blob/master/targets/m100/m100.asm

Hopefully this helps narrow the process of identifying the EU delta, which might be small.

**Comment by bkw777** [comment]
Another option in the mean time is, someone has gone the other way and patched the US M10 main rom with a German keyboard map.

This would require burning the main rom image to a 27C256 and installing inside the machine, and would make your EU M10 essentially into a US M10, except with different keyboard and without the modem hardware that any US rom expects.

I have no idea how much software will actually work on that, but they claim it works with at least TEENY.M10 and the M10 Multiplan option rom. That means at least TEENY and BASIC both have no problems using the serial port, so the missing physical modem hardware apparently doesn't break the uart routines.

https://www.mail-archive.com/m100@lists.bitchin100.com/msg17974.html

**Comment by a-brandt** [comment]
This seems to be a nice solution.

**Comment by a-brandt** [comment]
Hello bkw777,

I installed the ROM form Georg Käter and could connect to `dlplus`.

First I tried a ROM wich is based on the US-M10-ROM and used teeny. Then I used a ROM based on the Tandy100-ROM with a TSDOS option ROM.
Both worked for me.

**Comment by bkw777** [comment]
What do you mean based on tandy 100? Do you mean the one Steve Adolph provides for using REX# in a US M10? I'm surprised that works on a non-US keyboard or non-US hardware. But it's good to know.

**Comment by a-brandt** [comment]
Georg Käter has made a patch for the german keyboard. And I thinks it's based on the work of Steve Adolph.

**Comment by bkw777** [comment]
I got that. What I'm asking is what you meant by "Then I used a ROM based on the Tandy100-ROM with a TSDOS option ROM." What rom was that? That sounds like the one Steve made for use with rex#.

**Comment by a-brandt** [comment]
I can only guess. Georg Käter sent me the ROMs.

**Comment by bkw777** [comment]
ah ok
